GREP-498: Controller Reconciliation Prometheus Metrics (#499)
ai-dynamo/grove: docs/proposals/498-controller-reconcile-metrics/README.md

The built-in controller-runtime metrics tell us **that** reconciliation is slow or erroring, but not in sufficient detail to answer the following aggregate, fleet-level questions:

1. **No conflict attribution** — Status updates via `r.client.Status().Update()` and `r.client.Status().Patch()` can fail with 409 Conflict when an external controller (e.g., Dynamo's DGD controller) concurrently modifies the same PodClique/PodCliqueSet objects. The built-in `controller_runtime_reconcile_total{result="error"}` counts all errors equally — we cannot see the conflict rate or which resource type is affected.

2. **No error categorization** — Errors are returned with rich type information (`k8serrors.IsConflict`, `IsNotFound`, `IsAlreadyExists`, `IsTimeout`, etc.), but the built-in metrics collapse all errors into a single `result="error"` label. We cannot distinguish transient errors (conflict, timeout) from persistent errors (forbidden, invalid) at the fleet level.

> Additionally, sub-operation timing within a single reconcile (i.e., "which step of *this* reconcile is slow?") is a **per-request** question best served by distributed tracing, not by aggregate histograms with function-name labels. This proposal deliberately scopes itself to the aggregate, fleet-level questions where Prometheus metrics are the right tool, and tracks sub-operation visibility as future work under tracing (see [Future Work](#per-reconcile-sub-operation-visibility-via-tracing)).

Reviewer: please remove this, it not required for the GREP

### Goals

* Add a custom metric to track 409 Conflict errors by target resource kind, complementing built-in error counters that do not distinguish conflict attribution.
* Add a custom metric to categorize reconciliation errors by Kubernetes error type, enabling triage of transient vs persistent failures at the fleet level.

> * Document how Grove's built-in controller-runtime metrics answer standard reconciliation questions, so custom metrics remain scoped to what's genuinely missing.

Reviewer: I am not sure this should be part of the goals, it more on why the current metric does not cover by the suggested one

> * Stay within the aggregate, fleet-level scope that Prometheus metrics are designed for; defer per-request sub-operation visibility to distributed tracing.

Reviewer: I would remove this, its no adding information to the goals just confusing for  the prev work done on the GREP in earlier iteration

### Non-Goals

* This proposal does NOT duplicate built-in controller-runtime metrics (`controller_runtime_reconcile_time_seconds`, `controller_runtime_reconcile_total`, `controller_runtime_active_workers`). These are already exposed by the metrics endpoint and should be used as-is for reconcile-level observability.
* This proposal does NOT add metrics to Grove's scheduler backends (these are external components Grove supports, not part of Grove itself); it covers only the Grove operator controllers.
* This proposal does NOT add per-request, sub-operation timing via Prometheus metrics. That question belongs to tracing and is tracked as Future Work.
* This proposal does NOT modify the Grove CRD API, reconciliation behavior, or any user-facing functionality beyond the `/metrics` endpoint.

> * This proposal does NOT specify alerting rules or dashboard JSON (these are consumer concerns, not operator-side concerns).

Reviewer: please  change the capital NOT to not

## Proposal

### Overview

> Introduce two custom Prometheus metrics exposed by all three Grove controllers (PodCliqueSet, PodClique, PodCliqueScalingGroup). Both are **aggregate, fleet-level** signals that complement the built-in controller-runtime metrics without duplicating them.

Reviewer: As you present it already, we dont need to say it all over again.
It clear to the reader the metric should not cover by the existing and be fleet level.
```suggestion
Introduce two custom Prometheus metrics exposed by all three Grove controllers (PodCliqueSet, PodClique, PodCliqueScalingGroup).
```

> Introduce two custom Prometheus metrics exposed by all three Grove controllers (PodCliqueSet, PodClique, PodCliqueScalingGroup). Both are **aggregate, fleet-level** signals that complement the built-in controller-runtime metrics without duplicating them.

Reviewer: Missing the new metrics you are introducing

> How controllers record these metrics (recording helpers, context propagation, registration) is an implementation detail of the follow-up PR. This proposal describes only the **metric surface** — names, labels, semantics — and the rationale for each.

Reviewer: you can remove this line as well

> ### Relationship to Built-in Metrics

Reviewer: This section does not belong to here, the location should be after the details where you convince the reader its not cover today

| Question | Metric | Source |
|----------|--------|--------|
| How long does reconciliation take? | `controller_runtime_reconcile_time_seconds` (Histogram) | built-in |
| How many reconciles have occurred? | `controller_runtime_reconcile_total` (Counter, labeled by `result`) | built-in |
| How many reconciles are in flight? | `controller_runtime_active_workers` (Gauge) | built-in |
| How deep is the work queue? | `workqueue_depth` (Gauge) | built-in |
| How long do items wait in queue? | `workqueue_queue_duration_seconds` (Histogram) | built-in |
| How long before a retry? | `workqueue_unfinished_work_seconds` (Gauge) | built-in |
| API client performance? | `rest_client_request_duration_seconds` (Histogram) | built-in |
| **Which resource type is the 409 contention hotspot?** | `grove_operator_status_update_conflict_total` | **new (this proposal)** |
| **What fraction of errors are transient vs persistent?** | `grove_operator_reconcile_errors_total` | **new (this proposal)** |

> ### User Stories

Reviewer: The purpose of GREP user stories is to help us understand the features and capabilities users need that are currently missing from the project. Right now, some of the user stories are too technical and focus heavily on specific metrics. They need to be more high-level, focusing on the actual use case and clearly explaining what the user is trying to achieve and what is currently missing.

#### Story 1: Diagnosing Slow Deployment at Scale

**As** a Grove operator
**When** a user reports that deploying a large inference graph (50+ PodCliques) takes significantly longer than expected
**I want** to query built-in metrics (`controller_runtime_reconcile_time_seconds`, `workqueue_depth`, `workqueue_queue_duration_seconds`) to see whether the bottleneck is reconcile duration, queue backlog, or queue-wait latency
**So that** I can determine whether to tune `MaxConcurrentReconciles`, optimize the reconcile path, or investigate upstream load.

> For drilling into *which step of a single reconcile* is slow, tracing (not this proposal) is the right tool.

Reviewer: This user story is answer entirely by existing metrics, how it helps the case of missing metrics ?

* Keep the metric surface bounded (≤ 33 series), aggregate-only, and free of high-cardinality labels.

### Non-Goals

* This proposal does not duplicate built-in controller-runtime metrics (`controller_runtime_reconcile_*`, `workqueue_*`, `rest_client_*`). These are already exposed by the metrics endpoint and should be used as-is for reconcile-level observability (see [Existing Signals and Gaps](#existing-signals-and-gaps)).
* This proposal does not add per-request, sub-operation visibility. That question belongs to distributed tracing and is tracked in [Future Work](#per-reconcile-sub-operation-visibility-via-tracing).
* This proposal does not introduce high-cardinality labels (no namespace, no resource name, no function name, no pod UID). Such labels would couple the metric surface to internal code structure and break silently on refactor.
* This proposal does not add metrics to Grove's scheduler backends (these are external components Grove supports, not part of the Grove operator itself); it covers only the Grove operator controllers.
* This proposal does not modify the Grove CRD API, reconciliation behavior, or any user-facing functionality beyond the `/metrics` endpoint.

> * This proposal does not specify alerting rules or dashboard JSON (these are consumer concerns, not operator-side concerns).

Reviewer: can you rephrese the non goals, its should state what will not cover
most of them are just what no to do

### Existing Signals and Gaps

Grove's controller-runtime base already covers a large portion of standard observability. This proposal does not duplicate any of it. The remaining gaps below are what motivate the two new metrics.

| Question | Today | Gap |
|----------|-------|-----|

> | Reconcile latency / queue backlog / active workers (capacity-shaped) | `controller_runtime_reconcile_time_seconds`, `workqueue_depth`, `controller_runtime_active_workers` | None — covered by built-ins. |

Reviewer: if it is covered why is it here

| Total API 409 traffic | `rest_client_requests_total{code="409"}` | The `url` label is the request path, not the Kubernetes Kind being mutated. The metric **cannot attribute** a conflict to a specific Grove controller, nor to the status update target kind. 409s on `Pods` issued by upstream operators are indistinguishable from 409s on `PodCliqueSet` issued by Grove. |
| Error class breakdown (transient vs persistent) | `controller_runtime_reconcile_total{result="error"}` | All errors are collapsed into a single `error` label; conflict / timeout / rate-limited cannot be distinguished from invalid / forbidden. |

The full list of relevant built-in metrics is reproduced in [Built-in controller-runtime Metrics Reference](#built-in-controller-runtime-metrics-reference) for convenience.

## Proposal

### Overview

Introduce two custom Prometheus metrics exposed by all three Grove controllers (PodCliqueSet, PodClique, PodCliqueScalingGroup):

| Metric | Labels | Purpose |
|--------|--------|---------|

> | `grove_operator_status_update_conflict_total` | `controller`, `target_kind` | Attribute 409 conflicts to the contended resource type |

Reviewer: is it possible to get this information from the go client metric ? 

```suggestion
| `grove_operator_client_return_status_total` | `controller`, `target_kind` | Attribute 409 conflicts to the contended resource type |
```

Reviewer: lets keep return code as label

> | `grove_operator_reconcile_errors_total` | `controller`, `error_type` | Categorize errors by Kubernetes API error class |

Reviewer: What if the reconciled did not retrun k8s api error? does it mean every error return have to be typed ?

Both are aggregate, fleet-level signals that complement the built-in controller-runtime metrics. Cardinality bound, semantics, and recording verbs are covered in [Design Details](#design-details).

### User Stories

#### Story 1: Identifying the Cross-Controller Contention Hotspot

**Persona:** Grove operator on call.

**Scenario:** During a large inference graph rollout (50+ PodCliques), Pod scheduling latency spikes. I suspect status-write contention between Grove controllers and an upstream operator (e.g., NVIDIA Dynamo's DGD controller managing the same PodClique objects), but I have no way to confirm which Grove resource type is the hotspot.

> **Today's pain:** I have to grep controller logs across all three Grove controllers for "Conflict", manually correlate them by timestamp, and reason about which Grove resource type is most affected. This does not scale, cannot be alerted on, and produces no historical record I can plot or compare across releases.

Reviewer: I am not sure how the suggested metic will help with that, pod scheduling most of the time is scheduler related issue. even if you see conflict metric are high on spefic resource its defently means something is not right but here is not full correlation  with Pod scheduling latency

Reviewer: I would say that if you have expirnce such a conflict we might have other issue we need to solve

   **Mitigation:** All custom metrics use the `grove_operator_` namespace prefix, following Prometheus naming conventions and matching Grove's existing label (`app.kubernetes.io/name=grove-operator`).

3. **Risk:** Overlap with controller-runtime built-ins.
   **Mitigation:** This proposal explicitly avoids duplicating `controller_runtime_reconcile_*` and `workqueue_*` metrics. The [Relationship to Built-in Metrics](#relationship-to-built-in-metrics) table defines the boundary.

4. **Risk:** Coupling metric labels to internal code structure.
   **Mitigation:** No label value in this proposal references a Go function name, file path, or package. All labels are domain-level identifiers that survive refactors (`target_kind` = Kubernetes Kind; `error_type` = semantic class from `k8serrors.Is*`).

## Design Details

### Metric Surface

Two custom metrics are exposed. Both use the `grove_operator` namespace prefix and are registered against `ctrlmetrics.Registry` so they appear at the same `/metrics` endpoint as the built-in controller-runtime metrics.

#### `grove_operator_status_update_conflict_total`

- **Type:** Counter
- **Labels:** `controller`, `target_kind`

> - **Description:** Count of 409 Conflict errors returned by `Status().Update()` or `Status().Patch()` calls, attributed to the target resource kind.

Reviewer: why only status ? and not for spec / metadata

### Cardinality Bound

With three controllers, eight error types, and three target kinds, the total new time-series count is bounded by:

- `status_update_conflict_total`: 3 controllers × 3 target kinds = **9 series**
- `reconcile_errors_total`: 3 controllers × 8 error types = **24 series**

**Total: ≤ 33 new series.** No high-cardinality labels are introduced (no namespace, no resource name, no function name, no pod UID). The bound is static and refactor-insensitive.

### Recording Verbs (Contract Only)

Recording is intended to be synchronous, O(1), and side-effect-free on the reconciler's control flow:

- `status_update_conflict_total` is incremented in the error path of `Status().Update()` / `Status().Patch()` immediately after a 409 is detected (via `k8serrors.IsConflict`). The increment does not alter the returned error.
- `reconcile_errors_total` is incremented in the reconcile-loop error path, after the error is classified by the `k8serrors.Is*` predicates. A single error increments exactly one series.

Exact helper types, function signatures, registration code, and call sites are defined in the implementation PR and are not part of the contract reviewed here.

> ### Relationship to Built-in Metrics

Reviewer: this section should only talk about metic that are closed to the suggested metric but not cover what you ask for.

 
```suggestion
### Relationship to Existing Metrics
```

controller-runtime automatically registers a rich set of metrics via `sigs.k8s.io/controller-runtime/pkg/metrics`. The two custom metrics in this proposal sit alongside those built-ins; they do not replace them. The following table maps common reconciliation questions to the metric that answers them.

| Question | Metric | Source |
|----------|--------|--------|
| How long does reconciliation take? | `controller_runtime_reconcile_time_seconds` (Histogram) | built-in |
| How many reconciles have occurred? | `controller_runtime_reconcile_total` (Counter, labeled by `result`) | built-in |
| How many reconciles are in flight? | `controller_runtime_active_workers` (Gauge) | built-in |
| How deep is the work queue? | `workqueue_depth` (Gauge) | built-in |
| How long do items wait in queue? | `workqueue_queue_duration_seconds` (Histogram) | built-in |
| How long before a retry? | `workqueue_unfinished_work_seconds` (Gauge) | built-in |
| API client performance? | `rest_client_request_duration_seconds` (Histogram) | built-in |
| Which resource type is the 409 contention hotspot? | `grove_operator_status_update_conflict_total` | **new (this proposal)** |

> | What fraction of errors are transient vs persistent? | `grove_operator_reconcile_errors_total` | **new (this proposal)** |

Reviewer: redundant, we should only explain if there is similar metric that might paritly cover your use case but not all case

3. **No error categorization** — The built-in metrics only distinguish `success`, `error`, `requeue`, and `requeue_after`. We cannot distinguish a transient 409 Conflict (likely to self-resolve on retry) from a persistent validation error or a timeout, making it impossible to prioritize debugging effort.

### Goals

* Add three custom Prometheus metrics to fill gaps not covered by controller-runtime built-in metrics
* Track sub-operation timing within each reconcile loop to identify internal bottlenecks
* Track 409 Conflict errors with target-resource-kind attribution to diagnose cross-controller contention
* Track error categorization by Kubernetes error type for targeted debugging
* Document the full set of available metrics (built-in + custom) for Grove operators
* Design the framework as extensible — controller-specific business metrics can be added incrementally

### Non-Goals

* This proposal does NOT duplicate any controller-runtime built-in metrics

> * This proposal does NOT add metrics to the Grove scheduler component (only the operator)

Reviewer 2: There's no such thing as "Grove scheduler component". Grove supports different scheduler backends, none of them is part of Grove itself.

Author: Good catch — fixed. The Motivation/Non-Goals now read:

> *"This proposal does NOT add metrics to Grove's scheduler backends (these are external components Grove supports, not part of Grove itself); it covers only the Grove operator controllers."*

(commit 27c59ff, L83 in the revised README)

        return "forbidden"
    case k8serrors.IsInvalid(err):
        return "invalid"
    case k8serrors.IsTooManyRequests(err):
        return "rate_limited"
    default:
        return "internal"
    }
}
```

### Usage in Controllers

No `ObservedReconciler` wrapper is needed. Controllers create a `MetricsContext` at the beginning of their `Reconcile()` method:

```go
func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
    logger := ctrllogger.FromContext(ctx).WithName(controllerName)

    mc := metrics.NewMetricsContext(controllerName)
    ctx = metrics.WithMetricsContext(ctx, mc)

    // ... existing reconcile logic unchanged ...
}
```

Sub-operations are timed at step boundaries in `reconcileSpec` and `reconcileStatus`:

```go
func (r *Reconciler) reconcileSpec(ctx context.Context, logger logr.Logger, pclq *grovecorev1alpha1.PodClique) ctrlcommon.ReconcileStepResult {
    mc := metrics.MetricsContextFromContext(ctx)

    start := time.Now()
    if stepResult := r.ensureFinalizer(ctx, logger, pclq); ctrlcommon.ShortCircuitReconcileFlow(stepResult) {
        return r.recordIncompleteReconcile(ctx, logger, pclq, &stepResult)
    }

>     mc.RecordSubOperation("ensure_finalizer", start)

Reviewer 2: The interface looks good, we'll need to see how it lives together with the existing time measurements for scale-test milestone logging. Worth reviewing the existing code again (it might have changes since this PR was first drafted) and revisit the details.

Author: Good question — added a coexistence check to the Test Plan (commit 27c59ff):

> *"Verify these metrics do not duplicate or interfere with the scale-test milestone instrumentation in `operator/e2e/measurement/`. The new metrics expose aggregate per-controller signals via the `/metrics` endpoint; the measurement package continues to own scale-test milestone logging. Different pipelines, no expected interference."*

The two paths address different questions and emit through different sinks:
- **This proposal** → fleet-level aggregate metrics on the `/metrics` endpoint (Prometheus scrape), answering *"across all instances and time, how often does X happen?"*
- **`operator/e2e/measurement/`** → scale-test milestone logging for E2E test assertions, answering *"did this specific E2E run hit milestone Y within timeout Z?"*

The implementation PR will include an integration test asserting both are usable from the same operator binary without collision.

### Alt 3: ObservedReconciler Wrapper for All Metrics

**Pros:** Single wrapper provides all metrics automatically without per-controller changes.

**Cons:** Duplicates controller-runtime built-in metrics (`controller_runtime_reconcile_time_seconds`, `controller_runtime_reconcile_total`, `controller_runtime_active_workers`). Creates confusion with two sources of truth for the same data.

**Rejected because:** controller-runtime already provides reconcile-level metrics. Wrapping the reconciler would duplicate these. The `MetricsContext` approach focuses on what's genuinely missing (sub-operation timing, conflict attribution, error categorization) without duplication.

## Appendix

### Appendix A: Cross-Controller Conflict Storm Analysis

When Grove is used by an upstream operator (e.g., NVIDIA Dynamo), the following contention pattern occurs during large-scale deployments:

**Object ownership:**
```
Upstream Operator (e.g., Dynamo DGD Controller)
  creates → PodCliqueSet (PCS)
  scales  → PodClique (PC) / PodCliqueScalingGroup (PCSG) via Scale subresource
  reads   → PC/PCSG status for readiness checks

Grove Operator
  PCS Controller → creates PC/PCSG, updates PCS status

>   PC Controller  → creates Pods, updates PC status (replicas, readyReplicas)

Reviewer 2: Please use PCLQ instead of PC, across the whole document.

Author: Done — PC → PCLQ across the whole document, with PCS/PCLQ/PCSG formally defined in the new Appendix B terminology table (commit 27c59ff).
